macros/generic_tests/recency.sql

```diff
@@ -4,7 +4,19 @@
 
 {% macro default__test_recency(model, field, datepart, interval, group_by_columns) %}
 
-{% set threshold = dbt.dateadd(datepart, interval * -1, current_timestamp_backcompat()) %}
+{#- Get columns in model #}
+{%- set columns = adapter.get_columns_in_relation(model) %}
+
+{#- Get column type of field in model #}
+{%- set column_type = columns | selectattr("name", "==", field) | map(attribute="data_type") | first %}
+
+{#- Use a date-based threshold if column type is DATE or datepart is day or greater #}
+{%- if column_type == 'DATE' and datepart in ['year', 'quarter', 'month', 'week', 'day'] %}
```

Comment:
The comment and the code disagree here (one says and, the other says or) - which one did you intend to implement? 

Regardless, can you say more about why the checking of the datepart is even necessary? The original issue (#414) only talks about issues caused by the datatype of the column being date-only. Dropping the time components for _all_ thresholds will return inaccurate results. Consider this situation: 

```yaml
      - dbt_utils.recency:
          datepart: day
          field: my_datetime_column
          interval: 1
```

As I write this, it's `2022-11-23T04:57:19Z`. By dropping the time component even when it's available, the threshold will be 2022-11-22T00:00:00Z instead of T04:57:19Z, so the validity window is 5 hours larger, which would lead to false negatives. For example a record created at 1am would be considered valid even though it's actually 4 hours too old. 

```suggestion
{#- Use a date-based threshold if column type is DATE #}
{%- if column_type == 'DATE' %}
```
Let me know if I'm missing something!